Allow PHP 8.5 Support

sandesh-as opened the pull request:
<!--
Fill in the relevant information below to help triage your issue.

Assuming the current release is 1.5.0, the next patch release is 1.5.1, the next minor is 1.6.0 and the next major is 2.0.0, The Current release branch will be `1.5.x`, the next minor branch will be `1.6.x`, and the next major branch will be `2.0.x`

Pick the target branch based on the following criteria:
  * Documentation improvement: Current release branch 1.5.x
  * Bugfix: Current release branch 1.5.x
  * QA improvement (additional tests, CS fixes, etc.) that does not change code
    behavior: Next minor 1.6.x
  * New feature, or refactor of existing code: Next minor 1.6.x
  * Backwards incompatible features and refactoring: Next major 2.0.x

You MUST provide a signoff in your commits for us to be able to accept your
patch; you can do this by providing either the --signoff or -s flag when using
"git commit". Please see the project contributing guide and
https://developercertificate.org for details.
-->

|    Q          |   A
|-------------- | ------
| Documentation | yes/no
| Bugfix        | yes/no
| BC Break      | yes/no
| New Feature   | yes/no
| RFC           | yes/no
| QA            | yes/no

### Description

<!--

Tell us about why this change is necessary:
- Are you fixing a bug or providing a failing unit test to demonstrate a bug?
  - How do you reproduce it?
  - What did you expect to happen?
  - What actually happened?
  - TARGET THE CURRENT RELEASE BRANCH

- Are you adding documentation?
  - TARGET THE CURRENT RELEASE BRANCH

- Are you providing a QA improvement (additional tests, CS fixes, etc.) that
  does not change behavior?
  - Explain why the changes are necessary
  - TARGET THE NEXT MINOR BRANCH

- Are you fixing a BC Break?
  - How do you reproduce it?
  - What was the previous behavior?
  - What is the current behavior?
  - TARGET THE CURRENT RELEASE BRANCH

- Are you adding something the library currently does not support?
  - Why should it be added?
  - What will it enable?
  - How will the code be used?
  - TARGET THE NEXT MINOR BRANCH OR THE NEXT MAJOR IF BC WILL BE BROKEN

- Are you refactoring code?
  - Why do you feel the refactor is necessary?
  - What types of refactoring are you doing?
  - TARGET THE NEXT MINOR BRANCH OR THE NEXT MAJOR IF BC WILL BE BROKEN
-->

sandesh-as committed:
Allow PHP 8.5 Support

Signed-off-by: sandesh.as <cod48668@adobe.com>

sandesh-as committed:
Allow PHP 8.5 Support

Signed-off-by: sandesh.as <cod48668@adobe.com>

sandesh-as committed:
Allow PHP 8.5 Support

Signed-off-by: sandesh.as <cod48668@adobe.com>

sandesh-as committed:
Allow PHP 8.5 Support

Signed-off-by: sandesh.as <cod48668@adobe.com>

froschdesign commented on test/CodeGenerator/AbstractInjectorTest.php:
```diff
@@ -20,6 +20,7 @@
 
 /**
  * @covers \Laminas\Di\CodeGenerator\AbstractInjector
+ * @final This class should not be extended and will be marked final in version 4.0
```
Tests can be declared final because they are not covered by backward compatibility.

sandesh-as replied:
@froschdesign Updated all the tests to use the final keyword. Is it the same for TestAsset classes also? Can we declare it with a final keyword

sandesh-as committed:
Allow PHP 8.5 Support

Signed-off-by: sandesh.as <cod48668@adobe.com>

sandesh-as committed:
Merge branch '3.16.x' into Php-8.5-support

Signed-off-by: sandesh-as <cod48668@adobe.com>

sandesh-as commented:
@froschdesign We've completed the work to add PHP 8.5 support and Updated the changes as suggested, and the PR is now ready for review and merge.
Could you please take a look and share your feedback or any suggestions for improvement?

Thank you for your support!

sandesh-as committed:
Allow Php 8.5 support

Signed-off-by: sandesh.as <cod48668@adobe.com>

sandesh-as committed:
Allow Php 8.5 support

Signed-off-by: sandesh.as <cod48668@adobe.com>

sandesh-as committed:
Allow Php 8.5 support

Signed-off-by: sandesh.as <cod48668@adobe.com>

sandesh-as committed:
Allow Php 8.5 support

Signed-off-by: sandesh.as <cod48668@adobe.com>

kaktusas2598 commented:
Is there any update on this? We are trying to upgrade our platform to support PHP 8.5 and are held back by this package.

froschdesign commented on .laminas-ci.json:
```diff
@@ -1,5 +1,5 @@
 {
     "ignore_php_platform_requirements": {
-        "8.4": true
+        "8.5": true
```
@tux-rampage 
It would be great if you could do a review and if the requirements did not have to be ignored.
Thanks in advance! 👍🏻

tux-rampage replied:
This was due to a conflict with laminas-service manager that refused to install with 8.4 at the time this was added. Not sure if this is still needed.
Edit: It could also have been an issue with PHPUnit, but this is upgraded in this PR. 

If the test suite runs fine, I'm good with removing it.

tux-rampage replied:
Seems to be working fine without ignoring the platform reqs. So this is solved.

tux-rampage reviewed (changes requested):
Can you remove the "ignore_php_platform_requirements" from `.laminas-ci.json` to see if the CI-Pipeline still runs without errors?

Since you updated PHPUnit, this should no longer be an issue.

sandesh-as committed:
Merge branch '3.16.x' into Php-8.5-support

sandesh-as committed:
Allow Php 8.5 support

Signed-off-by: sandesh.as <cod48668@adobe.com>

sandesh-as committed:
Allow Php 8.5 support

Signed-off-by: sandesh.as <cod48668@adobe.com>

sandesh-as committed:
Allow Php 8.5 support

Signed-off-by: sandesh.as <cod48668@adobe.com>

sandesh-as commented:
Hi @tux-rampage, I’ve removed ignore_php_platform_requirements from .laminas-ci.json and fixed a few errors, including:

```
src/Config.php:193:21: PossiblyNullArrayOffset: Cannot access value on variable $this->types using possibly null offset null|string (see https://psalm.dev/125)

LaminasTest\Di\CodeGenerator\FactoryGeneratorTest::testGenerateForClassWithoutParams failing due to a mismatch:

--- Expected
+++ Actual
-'...container, $name = null, array $options = null): \...'
+'...container, $name = null, ?array $options = null): \...'
```


Could you please help to review these changes?

tux-rampage commented on src/Config.php:
```diff
@@ -190,6 +190,10 @@ public function getTypePreference(string $type, ?string $contextClass = null): ?
      */
     private function getTypePreferenceForClass(string $type, ?string $context): ?string
     {
+        if ($context === null) {
```
The only usage of this private method is in get type preference and it is never called with `null` for `$context`. Therefore it makes more sense to make this parameter not nullable instead of adding this check.

sandesh-as replied:
Updated

tux-rampage reviewed (changes requested):
Thanks for your follow up and addressing the issues.
Just a minor change on the parameter nullability. Would you take a look please.

sandesh-as committed:
Allow Php 8.5 support

Signed-off-by: sandesh.as <cod48668@adobe.com>

glo71317 commented:
@tux-rampage I think requested changes already done, Do we have an estimated timeline for when this change will be available for use?

tux-rampage reviewed (approved):
LGTM
